microcosm_pubsub/dispatcher.py

```diff
@@ -81,6 +91,13 @@ def validate_ttl(self):
         if int(self.opaque[TTL_KEY]) <= 0:
             raise TTLExpired()
 
+    def validate_processing_limit(self, message):
+        if self.max_processing_attempts > 0 and message.approximate_receive_count > self.max_processing_attempts:
+            raise SkipMessage(
+                "Message exceeded maximum of {max} processing attempts. Skipping",
+                extra=dict(max=self.max_processing_attempts),
+            )
```

Comment:
This logic is _very_ similar to `validate_ttl` just above.

I was going to switch and use `X-Request-Ttl` to achieve this effect. However, my understanding is that once enqueued both the body and attributes (metadata) of an SQS message are immutable.
This means that to update `X-Request-Ttl` on the message we'd have to first delete it from the queue, then re-enqueue it with the updated attributes. That would be quite a change from the way we do things today (and it feels wrong).

Because of that:
- I went with the approximate receive count
- I'm wondering if `X-Request-Ttl` is usable at all - it is if the intent is for the value to be updated by a message producer (i.e. either a gateway or the service producing the message, when re-processing the same request); it's not if the intent is for it to be updated by a daemon.

Reply:
I think it's reasonable to split the cases; we're using TTL to "hops" a request id follows globally, not just in pubsub.

In your case, you are only looking at local pubsub behavior.